pkg/consuldp/lifecycle.go

```diff
@@ -0,0 +1,195 @@
+		Addr:    fmt.Sprintf("%s:%s", cdpLifecycleBindAddr, cdpLifecycleBindPort),
+		Handler: mux,
+	}
+
+	// Start the proxy lifecycle management server
+	go m.startLifecycleServer()
+
+	return nil
+}
+
+// startLifecycleServer starts the main proxy lifecycle management server that
+// exposes HTTP endpoints for proxy lifecycle control.
+func (m *lifecycleConfig) startLifecycleServer() {
+	m.logger.Info("starting proxy lifecycle management server", "address", m.lifecycleServer.Addr)
+	err := m.lifecycleServer.ListenAndServe()
+	if err != nil && err != http.ErrServerClosed {
+		m.logger.Error("failed to serve proxy lifecycle managerments requests", "error", err)
+		close(m.errorExitCh)
+	}
+}
+
+// stopLifecycleServer stops the consul dataplane proxy lifecycle server
+func (m *lifecycleConfig) stopLifecycleServer() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.running = false
+
+	if m.lifecycleServer != nil {
+		m.logger.Info("stopping the lifecycle management server")
+		err := m.lifecycleServer.Close()
+		if err != nil {
+			m.logger.Warn("error while closing lifecycle server", "error", err)
+			close(m.errorExitCh)
+		}
+	}
+}
+
+// lifecycleServerExited is used to signal that the lifecycle server
+// recieved a signal to initiate shutdown.
+func (m *lifecycleConfig) lifecycleServerExited() <-chan struct{} {
+	return m.errorExitCh
+}
+
+// gracefulShutdown blocks until shutdownGracePeriod seconds have elapsed, and, if
+// configured, will drain inbound connections to Envoy listeners during that time.
+func (m *lifecycleConfig) gracefulShutdown(rw http.ResponseWriter, _ *http.Request) {
+	m.logger.Info("initiating shutdown")
+
+	// Create a context that  will signal a cancel at the specified duration.
+	// TODO: should this use lifecycleManager ctx instead of context.Background?
+	timeout := time.Duration(m.shutdownGracePeriod) * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	m.logger.Info(fmt.Sprintf("waiting %d seconds before terminating dataplane proxy", m.shutdownGracePeriod))
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		// If shutdownDrainListeners enabled, initiatie graceful shutdown of Envoy.
+		// We want to start draining connections from inbound listeners if
+		// configured, but still allow outbound traffic until gracefulShutdownPeriod
+		// has elapsed to facilitate a graceful application shutdown.
+		if m.shutdownDrainListeners {
+			err := m.proxy.Drain()
+			if err != nil {
+				m.logger.Warn("error while draining Envoy listeners", "error", err)
+				close(m.errorExitCh)
+			}
+		}
+
+		// Block until context timeout has elapsed
+		<-ctx.Done()
+
+		// Finish graceful shutdown, quit Envoy proxy
+		m.logger.Info("shutdown grace period timeout reached")
+		err := m.proxy.Quit()
+		if err != nil {
+			m.logger.Warn("error while shutting down Envoy", "error", err)
+			close(m.errorExitCh)
+		}
+	}()
+
+	// Wait for context timeout to elapse
+	wg.Wait()
```

Comment:
This method is subsequently broken up a bit in https://github.com/hashicorp/consul-dataplane/pull/130 to allow calling the blocking function directly instead of only from the HTTP endpoint.

I don't want to change the ordering here for that reason, but would sticking the blocking call into a new goroutine make sense?
```
func (m *lifecycleConfig) gracefulShutdownHandler(rw http.ResponseWriter, _ *http.Request) {
	go m.gracefulShutdown()

	// Return HTTP 200 Success
	rw.WriteHeader(http.StatusOK)
}
```